packages/flutter/lib/src/widgets/text.dart

```diff
@@ -713,3 +721,548 @@ class Text extends StatelessWidget {
     }
   }
 }
+
+// In practice some selectables like widgetspan shift several pixels. So when
+// the vertical position diff is within the threshold, compare the horizontal
+// position to make the compareScreenOrder function more robust.
+const double _kSelectableVerticalComparingThreshold = 3.0;
+
+class _SelectableTextContainerDelegate extends MultiSelectableSelectionContainerDelegate {
+  _SelectableTextContainerDelegate(
+    GlobalKey textKey,
+  ) : _textKey = textKey;
+
+  final GlobalKey _textKey;
+  RenderParagraph get paragraph => _textKey.currentContext!.findRenderObject()! as RenderParagraph;
+
+  @override
+  SelectionResult handleSelectParagraph(SelectParagraphSelectionEvent event) {
+    final SelectionResult result = _handleSelectParagraph(event);
+    if (currentSelectionStartIndex != -1) {
+      _hasReceivedStartEvent.add(selectables[currentSelectionStartIndex]);
+    }
+    if (currentSelectionEndIndex != -1) {
+      _hasReceivedEndEvent.add(selectables[currentSelectionEndIndex]);
+    }
+    _updateLastEdgeEventsFromGeometries();
+    return result;
+  }
+
+  SelectionResult _handleSelectParagraph(SelectParagraphSelectionEvent event) {
+    if (event.absorb) {
+      final SelectionEvent synthesizedEvent = SelectParagraphSelectionEvent(globalPosition: event.globalPosition, absorb: true);
+      for (int index = 0; index < selectables.length; index += 1) {
+        dispatchSelectionEventToChild(selectables[index], synthesizedEvent);
+      }
+      currentSelectionStartIndex = 0;
+      currentSelectionEndIndex = selectables.length - 1;
+      return SelectionResult.next;
+    }
+
+    // First pass, if the position is on a placeholder then dispatch the selection
+    // event to the [Selectable] at the location and terminate.
+    for (int index = 0; index < selectables.length; index += 1) {
+      bool globalRectsContainsPosition = false;
+      if (selectables[index].boundingBoxes.isNotEmpty) {
+        for (final Rect rect in selectables[index].boundingBoxes) {
+          final Rect globalRect = MatrixUtils.transformRect(selectables[index].getTransformTo(null), rect);
+          if (globalRect.contains(event.globalPosition)) {
+            globalRectsContainsPosition = true;
+            break;
+          }
+        }
+      }
+      if (globalRectsContainsPosition) {
+        if (paragraph.selectables != null && !paragraph.selectables!.contains(selectables[index])) {
```

Comment:
Yeah, i'll try to explain in depth. So say we have the following `Text` widget wrapped in a `SelectionArea`.

```dart
Text.rich( // SelectionContainer
  TextSpan(
    children: <InlineSpan>[
      TextSpan(text: 'This is some text before a widgetspan. '), // selectableA
      WidgetSpan(
        child: Text('This is some text in a widgetspan.'), // selectableC
      ), // SelectionContainer
      TextSpan(text: ' This is some text after a widgetspan.'), // selectableB
    ],
  ),
),
```

Internally the `RenderParagraph` of the root `Text` widget will create 2 selectables, one belonging to the text before the `WidgetSpan`, and the other belonging to the `Text` after the `WidgetSpan`. The root `RenderParagraph` will not create a selectable for the inline `WidgetSpan` because we explicitly skip placeholders when iterating through the text to create its selectables. These two selectables are registered to the `SelectionContainer` belonging to that root `Text` widget. The `RenderParagraph` belonging to the `Text` widget in the `WidgetSpan` will create one selectable fragment representing its entire text. This selectable fragment is then registered to the `SelectionContainer` belonging to the `WidgetSpan`, and that container registers itself to the container of the root `Text` widget above it. Then the root container registers itself to the root `SelectionArea`s registrar.

This is how it ends up breaking down:

```
SelectionArea.selectables = [ RootTextWidgetSelectionContainer ]
RootTextWidgetSelectionContainer.selectables = [ selectableA, WidgetSpanSelectionContainer, selectableB ]
WidgetSpanSelectionContainer.selectables = [ selectableC ]

RootRenderParagraph.selectables = [ selectableA, selectableB ]
WidgetSpanRenderParagraph.selectables = [ selectableC ]
```

I use this breakdown to determine if a given selectable is a placeholder (`WidgetSpan`), or if it belongs to the root text. Let me know if this was clear.

Reply:
Thanks for the thorough explanation. Where does paragraph.selectables get populated? I'm not seeing it at first glance.

Also, why didn't we need this infrastructure before this PR? It seems like you would need this for selecting by line etc. also.